### ISSUE TITLE
What html and css frameworks should we use (if any)?

### ISSUE THREAD

**Issue by Exellin** [emerald-rubies/tenacious#44]
This was mentioned in #32 and i figured it could use its own issue.

@pachonk brought up not wanting to go farther than something like bourbon.io which looks like a library to make writing scss easier, but doesn't actually have any built in html components or css classes.

I'm fine with going for vanilla, but I think its worth discussing unless alex wants to force it.

**Comment by pachonk** [comment]
It's @emeralds project, I'm just here to act in a "client/project owner" seat.

**Comment by sanchez2210** [comment]
I like Susy

**Comment by pachonk** [comment]
Is the feeling mutual?

**Comment by pachonk** [comment]
Sorry, that was a lame joke.
In all seriousness, I would LOVE for us to go light on all assets to make this load fast even on a 1x connection.

**Comment by sanchez2210** [comment]
Haha she was hard at first but I persisted. Now I include her in most of my projects

**Comment by pachonk** [comment]
I just laughed so hard my upstairs neighbor stomped and said "SHUT UP I'M TRYING TO WATCH A MOVIE!" omg :joy:

**Comment by Exellin** [comment]
If all we need is a grid I have access to a simple grid we can just put into a css file.

**Comment by pachonk** [comment]
Even a grid isn't fully needed with the idea of flexbox. I could be wrong
tho

**Comment by Exellin** [comment]
Yeah we could do flexbox Ive dabbled in it a bit and this project could be a good place to practice it.

**Comment by pachonk** [comment]
I was fortunate to dabble with flexbox on a project with @brandondees and it was TONS of fun! if you do it perfect, it can make a responsive design with almost no media queries.

**Comment by brandondees** [comment]
It might make sense to back up and think in terms of what the project's objectives and requirements are.... is the point to practice working with frameworks X, Y, and Z? then get clear on that as a team and pursue it. Is the point to focus on getting team collaboration skills up while working with familiar technologies? get clear about that priority. Is the point to simulate a real-world client-driven project? If so, what are your browser / device support requirements likely to be? What level of graphic/UX design effort should go into the initial planning stages vs delivering functional-only prototypes? @pachonk and I can play the role of product owners if y'all want to do it that way, but satisfying our expectations will require planning and team focus on the deliverable priorities we establish, as opposed to "work on whatever part"

**Comment by Exellin** [comment]
My Impression for this project was that is is a way for us to practice collaboration in a project through code reviewed pull requests.  This improves our skills as developers as all of our code is reviewed before it is accepted and gives us an idea of best practices followed by professionals in the field.

I don't think the focus of this project should be on the graphics to begin with (as in a pull request shouldn't be denied because the elements are not styled sufficiently).  Since additional styling can be done at a later date.  However the user experience should be as close as possible to the mvp for each feature (as in deciding whether to redirect to a new page to edit an element or turn the view in an edit form with javascript/ajax.  Also the general layout of a page).  Since changing the ux is much more time intensive at a later date and disrupts the flow of user stories.

I think for this first project the barrier of entry to contribute should be as low as possible.  We shouldn't have a requirement of people having to spend hours/days acquainting themselves with a new framework in order to contribute.  However this isn't necessarily vanilla, as I think plenty of developers (myself included which was a mistake) jump straight into bootstrap and don't know how to work with vanilla html and css to create standard web elements.

I personally would like to create this project using flexbox (as I know the basics but haven't created a project with it yet, plus we have awesome mentors to guide us along the way.  However I'm not sure if everyone feels the same way about a new technology and it contradicts my point above about the 
barrier of entry being as low as possible).

So far this project has had very minimal actual planning and laying out of requirements (I assume in the real world this process takes weeks/months before any actual code is written, correct me if I'm wrong here).  So far I am enjoying the "create issues and work on whatever part" development style as it allows us to look through the issues list and pick whatever we would like to work on when we have the time.

If we were to go the route of Brandon and Alex becoming product owners with requirements, I think that would require a overhaul of how we are approaching this project and I think it would be best left to a project at a later date.

I would love to hear everyone elses opinion on these topics.

**Comment by pachonk** [comment]
> I don't think the focus of this project should be on the graphics to begin with (as in a pull request shouldn't be denied because the elements are not styled sufficiently).

If a pull request is sent in _with the purpose of fixing a UI glitch_ and the glitch still exists, it'll be unlikely for me to pull it in

> However this isn't necessarily vanilla, as I think plenty of developers (myself included which was a mistake) jump straight into bootstrap and don't know how to work with vanilla html and css to create standard web elements.

There are enough BRAND NEW DEVS that I'm willing to say it's worth learning "the vanilla way". Also, it's worth learning yourself the vanilla way. As we move forward as a group of devs, maybe we can pick out some libraries, but I think this particular point you made is "all the more reason to go vanilla"

> plus we have awesome mentors to guide us along the way.

:blush: 

> However I'm not sure if everyone feels the same way about a new technology and it contradicts my point above about the barrier of entry being as low as possible

I'm not gong to lie, I find flexbox easier than normal css positioning. That said, it's not a fix all solution

> If we were to go the route of Brandon and Alex becoming product owners with requirements, I think that would require a overhaul of how we are approaching this project and I think it would be best left to a project at a later date.

For this first project, we'll be rather helpful towards the overall project. The reason the planning phase was is minimal is because I heard a consensus that we wanted to CODE CODE CODE. @brandondees and I are already discussing the style we want to handle the next project and we believe that we can simulate a "freelance client" PERFECT for practice on taking in client requests and executing them.

**Comment by Exellin** [comment]
Thanks for the in depth reply Alex.  You brought up some great points especially about going for vanilla in order for new developers to learn how to create a project without having to use something like bootstrap as a crutch even if they are more familiar with it.

If we aren't using a framework or grid system (which you just convinced me is the best path going forward) we will have to use multiple css tools in order to style and layout our document, and flexbox is just one of those tools that I would love to learn and work on in this project.

So my vote is: Vanilla with flexbox.

**Comment by pachonk** [comment]
@Exellin flexbox is just a css property. Realistically it doesn't HAVE TO BE USED on all parts of the app, so your vote is _really_ "Vanilla" and you intend to use flexbox during your development phase of the frontend :)

**Comment by brandondees** [comment]
> Since changing the ux is much more time intensive at a later date and disrupts the flow of user stories.

I'll contend that UX is much more streamlined when user-focused stories are the driving force behind defining any requirements, and should usually require no major changes. Major changes that are needed should be discovered as early as possible through a feedback driven short-iteration design and dev process, before additional interdependencies have time to form. Needing to change the UX in a major way after the fact is generally a sign that user needs weren't considered when implementing it the first time, which is commonly a side effect of emphasizing the tools and technology and what's easy for the developers rather than finding the shortest path to user happiness. The classic example of this is something like ActiveAdmin for generating an essentially "free" full database CRUD admin interface. Yep, it technically makes it possible for an admin user to do whatever needs to happen with the data in the DB, but generally it's very awkward workflow-wise to use for repetitive cross-model workflows without very heavily modifying the default CRUD arrangement, at which point it becomes an obviously more cost-effective approach to just use rails itself to create task-oriented (rather than data oriented) interfaces for admin workflows. "find a user account by email and give them a free month of service" is a text field with autocomplete and a "extend service 1mo" button if you design a task-oriented UI for it (this is maybe a 15 minute feature for an experienced rails dev), whereas taking full advantage of the scaffolded CRUD UI  is a 5 minute feature and has a lot more capabilities as well, but the admin user will have to dig around for more than 15 minutes each time to find the user account based on email and then trace that to the CRUD UI for the account subscription expiration status field and then update it.

> So far this project has had very minimal actual planning and laying out of requirements (I assume in the real world this process takes weeks/months before any actual code is written, correct me if I'm wrong here).

Yeah no real time has been put into that yet, that's for sure. It doesn't need to take weeks/months, especially given @pachonk  already has a good idea of the functionality expected, but for any given project it is likely to be reasonable to spend a number of intensive days or possibly weeks on pre-development work of this kind. Given nobody here is working on this full time to the best of my understanding, we can try to 80/20 that into a subset of features and then make requirements for additional phases of development once we get a core proof of concept project delivered. It's also possible to have this requirements/planning work happening in parallel with initial development, which I think is applicable here. In an agile process, the goal is generally to ship _some kind of working software_ by the end of each iteration, which are often some recurring length of time between 2-6 weeks. For a small / lightweight project like this I think 1 or 2 weeks should be a reasonable iteration length.

**Comment by pachonk** [comment]
What do you mean 80/20?

**Comment by troystarwalt** [comment]
I love Flexbox and the support is great across the browsers that we are targeting. So, plus one for no frontend framework.

**Comment by brandondees** [comment]
https://en.wikipedia.org/wiki/Pareto_principle AKA the 80/20 rule, very often referenced in the context of prioritizing design and development tasks.

**Comment by brandondees** [comment]
experimental poll:
(click choice to vote) 

[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/bootstrap)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/bootstrap/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/foundation)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/foundation/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/bourbon%20(%2Bneat%2C%20bitters%2C%20etc%20potentially))](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/bourbon%20(%2Bneat%2C%20bitters%2C%20etc%20potentially)/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/vanilla%20CSS)](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/vanilla%20CSS/vote)
[![](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/other%20(add%20suggestion%20comments))](https://m131jyck4m.execute-api.us-west-2.amazonaws.com/prod/poll/01BMBA1WDF9JEAQATF0S0TVBTD/other%20(add%20suggestion%20comments)/vote)

**Comment by Talha5** [comment]
hehe, even one can vote again on the same poll.

**Comment by pachonk** [comment]
I couldn't!

**Comment by Talha5** [comment]
I voted the other day, and then i did it just now. :|

**Comment by pachonk** [comment]
Must have 2 different IP addresses then.

**Comment by pachonk** [comment]
Closing this now. We can open another more specific issue later.
